# Homework 5: Loops and conditionals

## Submission instructions

Upload the `ipynb` file to Canvas:
> File -> Download -> ipynb -> upload to Canvas (like any other file)

*Only* the `ipynb` file type will be accepted.

Save a copy of the notebook right away to avoid losing your work!

## Data file for this assignment

Run this code cell to load the same periodic table properties table from Lectures 9, 11, and 12.

## Dataset: Periodic Table Properties

This dataset contains basic physical and chemical properties of chemical elements
from the periodic table, including standard atomic weight, density, atomic radius,
electronegativity, melting point, and crystal structure.

Columns include: `Element`, `Standard Atomic Weight`, `Density (g/cm^3)`,
`Atomic Radius`, `Electronegativity`, `Melting Point`, `Structure`, and others.

In [1]:
import os
import pandas as pd

_file = 'periodic_table.csv'
_github = f'https://raw.githubusercontent.com/wfreinhart/matse219/main/datasets/{_file}'
_local = next(
    (p for p in (f'datasets/{_file}', f'../datasets/{_file}', f'../../datasets/{_file}', f'../../../datasets/{_file}')
     if os.path.exists(p)),
    None,
)

try:
    data = pd.read_csv(_local if _local else _github)
except Exception as e:
    raise RuntimeError(
        f"Could not load '{_file}'. If you are in Colab, check your internet "
        f"connection and that the file exists at {_github}"
    ) from e
data

,Element,Standard Atomic Weight,Atomic Radius,Structure,Electronegativity,Valence,Melting Point (K),Density (g/cm^3),Heat of Fusion (kJ/mol),Molar Heat Capacity (J/mol*K),Young's Modulus (GPa),Poisson Ratio
0,Cu,63.5460,0.128 nm,FCC,1.8,2+,1357.7700,8.960,13.260,24.440,119.0,0.340
1,Ni,58.6934,0.125 nm,FCC,1.8,2+,1728.0000,8.908,17.480,26.070,200.0,0.310
2,Al,26.9810,0.143 nm,FCC,1.5,3+,933.4700,2.700,10.710,24.200,70.0,0.350
3,Zn,65.3800,0.133 nm,HCP,1.7,2+,692.6800,7.140,7.320,25.470,108.0,0.250
4,Pb,207.2000,0.175 nm,FCC,1.6,2+,600.6100,11.340,4.770,26.650,16.0,0.440
5,Sc,44.9550,0.162 nm,HCP,1.4,2+,1814.0000,2.985,14.100,25.520,74.4,0.279
6,Ti,47.8670,0.147 nm,HCP,1.5,4+,1941.0000,4.506,14.150,25.060,116.0,0.320
7,V,50.9415,0.134 nm,BCC,1.6,5+,2183.0000,6.110,21.500,24.890,128.0,0.370
8,Cr,51.9961,0.128 nm,BCC,1.7,6+,2180.0000,7.190,21.000,23.350,279.0,0.210
9,Ga,69.7230,0.135 nm,ORTHO,1.8,3+,302.9146,5.910,5.590,25.860,9.8,0.470


# Problem 0 (0 pts): generative AI usage statement

As you work on this assignment, feel free to use generative AI tools to help you learn,
understand, and debug Python code. In particular, you could get hints or conceptual guidance in
the implementation you write yourself.

However, you must clearly disclose and cite all use of AI. You must include:
1. The name(s) of the AI tool(s) used.
2. The specific prompt(s) you used to generate the content.
3. A description of how you used the output and what edits or additions you made to integrate it
   into your own work.

You are fully responsible for the final submitted work -- critically evaluate, fact-check, and
verify all AI-generated content for validity. Failure to properly cite and disclose AI use
constitutes plagiarism under Penn State's Academic Integrity policy.

Write your disclosure (or "I did not use an AI tool for this assignment") in the cell below.

*your disclosure here*

# Problem 1 (20 pts): Loop basics -- rows and accumulators

(a) Build a dict called `radius_lookup` that maps each `Element` to its `Atomic Radius`, using
`zip` and `dict` (same pattern as `density_lookup` in Lecture 12). Note that `Atomic Radius` is
stored as text (e.g. `'0.128 nm'`), not a number -- that's fine, you're just looking it up, not
doing arithmetic on it.

(b) Loop over `radius_lookup.items()` and print each element together with its radius.

(c) Using `.iterrows()`, loop over `data` and build a Python list called `reports`, where each
entry is a formatted string of the form `"Cu (FCC): 8.96 g/cm^3"` -- element symbol, structure in
parentheses, then density. (Hint: an f-string with `:.2f` on the density will keep the formatting
clean.) Display `reports` when you're done.

# Problem 2 (25 pts): Loop over columns -- a summary stats table

(a) Not every column in `data` is actually numeric, even if it looks like it should be. Print
`data.dtypes` and identify, in a comment, which two columns store numbers as text rather than as
a numeric dtype, and briefly say why (one sentence each).

(b) Build a Python list called `numeric_cols` containing the names of every column that
genuinely is numeric (`dtype` of `float64` or `int64`) -- don't just copy the list from lecture,
write the check yourself (you may do this by hand, referencing part (a), or by filtering
`data.dtypes` programmatically).

(c) Loop over `numeric_cols` and build a dict-of-dicts called `summary`, where each entry stores
`'mean'`, `'std'`, `'min'`, `'max'`, **and** `'range'` for that column, where `range` is `max -
min`. (`range` is new -- it wasn't part of the Lecture 12 worked example, so you'll need to add
it to the loop body yourself.)

(d) Convert `summary` to a `DataFrame` the same way Lecture 12 did (build it, then `.T`) and
display it. Which numeric column has the largest `range` relative to its `mean`?

# Problem 3 (25 pts): `if`/`elif`/`else` -- a three-way classifier

(a) Bare-Python warm-up, no loop yet: given a single value `hf = 15.0` (kJ/mol), write an
`if`/`elif`/`else` chain that assigns `label` as `"low"` if `hf < 8`, `"medium"` if `8 <= hf <
16`, or `"high"` otherwise, and print `label`.

(b) Now loop over `data['Heat of Fusion (kJ/mol)']` with the same thresholds as part (a), building
an accumulator list called `fusion_labels`, and print each value alongside its label as you go.

(c) In a markdown cell (or a comment), explain in 1-2 sentences why this three-way labeling could
not be done with a single vectorized comparison like `data['Heat of Fusion (kJ/mol)'] > 16` the
way Lecture 11 filtered rows.

# Problem 4 (30 pts): Loop + conditional -- a new categorical column

(a) Write a loop over `data['Density (g/cm^3)']` that classifies each element into a weight class
using `if`/`elif`/`else`: `"light"` if density < 5, `"medium"` if 5 <= density < 8, `"heavy"`
otherwise. Accumulate the labels into a list and assign it to a new column, `data['Weight
Class']`.

(b) Display `data[['Element', 'Density (g/cm^3)', 'Weight Class']]` for all 14 rows and check
your loop's output by eye against the density values.

(c) Without using `groupby` (you haven't met it yet -- that's a later lecture), use a loop and a
plain dict accumulator to count how many elements fall into each weight class, and print the
result. (Hint: initialize a dict like `counts = {'light': 0, 'medium': 0, 'heavy': 0}`, then loop
over `data['Weight Class']` and increment the right entry each time.)

(d) In a comment, name one column *other than* `Density` or `Young's Modulus` that you think
would make a materially meaningful three-way category, and what your thresholds would be. (No
need to implement it -- just the column and thresholds.)

# Wrap-up

This homework only asked for `for` loops, `dict` accumulators, and `if`/`elif`/`else` chains --
the same two patterns from Lecture 12 (loop over column names to avoid repetition; loop +
conditional over rows to build a new label), applied to different columns and thresholds than the
lecture used. If you found yourself reaching for `data['col'] > value` to build a multi-way label
in one line, that's a sign to stop and use the loop instead -- that's exactly the judgment call
this homework is testing.